Source/Dafny/AST/DafnyAst.cs

```diff
@@ -6142,6 +6210,51 @@ public BoundVar(IToken tok, string name, Type type)
       Contract.Requires(name != null);
       Contract.Requires(type != null);
     }
+
+    public bool CurrentTypeAssumedToBeCompilable { get; set; } = false;
+
+    public void AssumeCompilableType(Type secondaryType) {
+      Contract.Assert(!CurrentTypeAssumedToBeCompilable);
+      otherType = secondaryType;
+      AssumeCompilableTypeIfAny();
+    }
+
+    public void AssumeCompilableTypeIfAny() {
+      if (!CurrentTypeAssumedToBeCompilable && otherType != null) {
+        (otherType, type) = (type, otherType);
+        CurrentTypeAssumedToBeCompilable = true;
+      }
+    }
+
+    public void AssumeOriginalType() {
+      if (CurrentTypeAssumedToBeCompilable) {
+        Contract.Assert(otherType != null);
+        (type, otherType) = (otherType, type);
+        CurrentTypeAssumedToBeCompilable = false;
+      }
+    }
+
+    public void AcceptOriginalTypeAssumption() {
+      if (CurrentTypeAssumedToBeCompilable) {
+        Contract.Assert(otherType != null);
+        (type, otherType) = (otherType, null);
+        CurrentTypeAssumedToBeCompilable = false;
+      } else {
+        otherType = null;
+      }
+    }
+
+    public Type CompilableType => CurrentTypeAssumedToBeCompilable ? Type : otherType;
+    public Type OriginalType => CurrentTypeAssumedToBeCompilable ? otherType : Type;
+
+    public bool NeedsToVerifyOriginalTypeInference() {
+      return otherType != null && !OriginalType.IsCompilable();
+    }
+    public bool CompilableTypeAssumptionIsNotTrivial() {
+      return otherType != null && !otherType.Equals(Type, true);
+    }
+
+    private Type otherType = null;
```

Comment:
Can you put fields at the top of the type declaration? Also, what is the meaning of otherType? Could you add a comment?